### ISSUE TITLE
How to compute perplexity after beam search with torch generator agent?

### ISSUE THREAD

**Issue by gabrielle-lau** [facebookresearch/ParlAI#3856]
Hi, I'd like to compute perplexity of a Blender 90M generative model after the word probabilities ("scores") in the [_generate function](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L1131-L1135) for beam search of size 3 have been artificially modified by a custom decoding algorithm. In the ParlAI code,  PPL is based on computing cross-entropy loss and [record_local_metric](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L706). Is there a way to tweak the [compute loss function](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L682) and call it later after beam search to compute the desired PPL? 

**Things I've tried**
- While my custom dataset does not contain labels, I added some "fake" labels (since I don't need to compute accuracy metrics), so that cross-entropy loss and thus PPL could be computed using the existing [PPLMetric](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L316). 
- Replace the [score variable in compute_loss](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L694) with the last time step's [modified score in _generate](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L1131), but the difference in input and target tensor sizes (due to beam size 3) caused a value error when computing [loss](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L696). 
- Compute PPL myself by summing up log likelihoods stored in beams[0].all_scores (inserted on [line 1163](https://github.com/facebookresearch/ParlAI/blob/dd4fd1f4e2de2cd2ec0d84f3f98bd3bf5bfa4b50/parlai/core/torch_generator_agent.py#L1163)), then average and take exponential. However, this yields incorrectly large PPL>100 compared to ParlAI's code that yields PPL=8.4 (when I don't modify word probabilities). The working ParlAI command is as follows:

```
parlai eval_model --task hostile_sexism -dp /project/my_data \
--model-file /project/checkpoint/Reddit_90M_genderation_LRx8.1.checkpoint \
--metrics ppl --beam-size 3 --inference beam
```

Appreciate any guidance! Thank you :)

**Comment by stephenroller** [comment]
Your last approach is the right approach, just a little off. I believe this is correct, computed here:

https://github.com/facebookresearch/ParlAI/blob/173bb9f1a7487588521c4ad7ee0832c6a5895425/parlai/core/torch_generator_agent.py#L880

```python
>>> [sum(PPLMetric.many([-score[0]*score[0].exp() for score in beam.all_scores]), None) for beam in beams]
... [PPLMetric(1.043)]
```

**Comment by gabrielle-lau** [comment]
Thank you! I used your approach and got a similar value. I guess it's normal for PPL to decrease a lot (from PPL=8 to 1) after beam search?

**Comment by stephenroller** [comment]
Not sure, first time I measured it. But we are asking the model to evaluate its most-confident answer, not a gold response, so a significant drop is to be expected.

**Comment by gabrielle-lau** [comment]
That makes sense, thank you! :)
